### ISSUE TITLE
Add support for advanced BBS features like pseudonyms

### ISSUE THREAD

**Issue by msporny** [w3c/vcalm#396]
Protocols, such as VC API, that support unlinkable disclosure schemes (such as BBS or CL Signatures) with features such as pseudonyms and binding to a secret that a holder possesses, need to provide mechanisms to achieve those features. The VC API should add fields (like prover ids and verifier ids and commitments) to the protocol (perhaps via vp-request-spec) and examples of how to achieve these advanced features. 

/cc @Wind4Greg

**Comment by msporny** [comment]
These features need to be incubated in the VC Playground and a wallet or two before we add the text to the specification.

**Comment by msporny** [comment]
The group discussed this on 2025-03-25:

@dlongley noted that DB has been doing some work in this area:

in a VC in a VP
  - cons:
    - new VC type
    - new context
    - does VC need a proof?
  - pros:
    - works with any protocol using VCs  
    - maybe leverage Enveloped VC w/o creating a translation at first?

in a VP directly
  - cons:
    - new VP type
    - new context
  - pros:
    - works with any protocol using VPs

in a VC API json payload in a new property
  - cons:
    - very custom, requires changes to VC API
  - pros:
    - no contexts required, specific to VC API

The current things you can send in VC API -- VP, VPR, redirect URL. We need to be able to put new information for pseudonyms -- either in one of those existing fields, or create a new field. Put something in existing field and it can be used more broadly. If we drop in VC API, not as reusable, but faster to get done (not as much re-use). @PatStLouis noted that this is relevant to BBS and Anoncreds: https://api.anoncreds.vc/docs -- commitment requirement -- step missing where holder needs to include blind claim -- sign blind commitment, what I landed on credential request endpoint, holder provides proof of commitment and that request is passed at issue endpoint in options as request proof -- don't know about other cryptosuites, but every blind claim that issuer doesn't know value (issuer blind signs) -- content of request proof depends on each cryptosuite (know what to do for anoncreds-bbs, but maybe not for another type of cryptosuite). @dlongley noted that that design doesn't seem to slot into exchange framework/protocol -- going out to another endpoint to do something new -- don't get benefits from exchange process. @PatStLouis noted that these new endpoints could be called as a part of an exchange -- holder provides sensitive info to endpoint, forward request to issuer -- way calls are being done are done through a workflow exchange. @dlongley noted that might be talking about two different things -- holder service is different piece of interaction, interaction between holder and exchange service between holder/issuer runs. Two things: 1) construction of credential request holder -> holder service, included in exchange (might need new issue to track this), 2) use of that information in an exchange process. @dlongley inside of an exchange, how do you express the information (this request?). @PatStLouis noted that how exchange happens depends on many things. Exchange could look very different, but in many cases, holder needs to give verifier blinded claims to include in credential signature. @dlongley That could happen through VC API service or directly through device. @PatStLouis in Anoncreds, wallet creates that locally... if we want that happen externally, then we could use a new VC API endpoint. 2nd part that's missing, presentation request -- is this issue just around issuance? What about presentation? Is QueryByExample sufficient? What about commitments, range proofs, verifiable encryption, etc? Advanced features that are unique to subset of cryptosuites are probably not simple enough for QBE -- how can different features be requested? @dlongley noted that we should look into that -- there is another query type for DID Auth, could be similar to that, but might be the case that for some credentials, there is no other way to do the presentation... or the way you do the presentation is you must construct a presentation that has a pseudonym that contains the domain, etc. You might not need an additional query type, receive a presentation request, we'd specify how that is done in spec. @PatStLouis Different pseudonyms can be revealed w/ different domains -- verifiable encryption can be used by encryption key from issuer for specific pseudonym to be encrypted, or you might want to do w/ verifier -- one domain for full presentation might not be suitable, might need domain per pseudonym claim. @dlongley suggested that default behavior would be domain doing presentation to, if that needs to be customized, we might do it in a way that we do DID Auth (wallets that don't understand could throw up error). We keep things simple for 80%/90% use case -- slots into how regular humans are used to interacting with websites and trust signals.

Next steps: Create two categories of use cases: issuance and presentation. For each category, break down the most common expected use cases from less common use cases.

**Comment by msporny** [comment]
The most common use case for issuance: Get a BBS credential issued that includes pseudonym support and no special fields (default behavior to get BBS VC that allows you to present with pseudonym).

Other use cases for issuance:

* Issuer requires the Holder to reveal their pseudonym secret so that presentations can be linked by the Issuer. The Holder still gets per-verifier unlinkability, but you don't get Issuer unlinkability with this approach.
* Issuer has specific fields that they will blind sign that the issuer doesn't see, but Holder wants to be able to present that the Issuer saw a "proof of existence" over some blind-signed data. (this is a stretch from a "commercial value" perspective).

The most common use case for presentation: Present a basic BBS credential that has pseudonym support to a website where the context used to generate the pseudonym is the web origin. Expectation is 90%+ of the use cases are this use case.
 
Other use cases for presentation:

**Comment by ottonomy** [comment]
@msporny I don't quite understand the "In a VP Directly" mechanism, like inside the existing `verifiableCredential` property, or in a different property added by an additional optional context used sometimes within a VP returned that includes something other than a credential?

I suspect that "in a VC in a VP" is the option that I prefer due to it being "the most commonly extended object"

**Comment by dlongley** [comment]
I think we need this in 1.0 so I've added that tag for the group to consider.

**Comment by kezike** [comment]
@PatStLouis I propose that we revisit this issue on the next call, so that we can make a decision ASAP.

**Comment by dlongley** [comment]
Some quick thoughts on this:

If we define ~`additionalData`~ `exchangeData` as a property on a VP (directly or with a new, additional VP type of say ~`AdditionalDataVerifiablePresentation`~ `ExchangeVerifiablePresentation`), we might use it like this:

1. In response to a VPR that includes an offer for VCs using BBS (see #689), the VP response could have an `exchangeData` field with a value of `{"bbs-2023": {/* necessary commitment information to create a pseudonym */}}`.
2. In response to a VPR that asks for zcaps, where the request includes a reference ID for each zcap to be generated, the VP response could have an `exchangeData` field with a value of `{"zcaps": {referenceIdFoo: <zcap>, referenceIdBar: <zcap>}}`.
3. If there is any additional information that should be applied when presenting payment information, for example when presenting a PaymentRequest VC and a PaymentMethod VC, an `exchangeData` field with a value of `{"payment": {/* authorized amounts, memos, etc. */}` could be used.

All of the information in `exchangeData` would be signed (if the VP is signed), as usual.

It's not clear that very use of this property would be protocol-specific, but it is expected to be "bilateral" or "peer-based", with an understanding that the information is not globally unambiguous. But we could also consider these names:

~`protocolData` (for the property) and `ProtocolVerifiablePresentation` for the type (if we add this as an extension instead of as a core VP property).~ Other names to consider might be `exchangeData` and `ExchangeVerifiablePresentation`.

I don't think the information is ever meant to be understood outside of a particular exchange, so the best names here might indeed be: `exchangeData` and `ExchangeVerifiablePresentation`.

**Comment by eric-schuh** [comment]
@dlongley I am currently working on a PR for this and wanted to ask if the ```exchangeData``` object needs any structured format. In the above discussion you had provided a few examples but I am not getting the sense that the form of this object should be expected to be well known as it could be any number of data types being communicated. 

My plan is to make the ```exchangeData``` object as generic as possible, then provide some examples of what the object may look like using bbs/zcaps/payments/etc as you listed above. 

Any issues with this plan?

**Comment by dlongley** [comment]
> [@dlongley](https://github.com/dlongley) I am currently working on a PR for this and wanted to ask if the `exchangeData` object needs any structured format.

No, I think it should specifically just be typed as `"type": "@json"` and carry any arbitrary JSON (to be determined by the exchange at hand and the specification that governs the protocols in use).

> My plan is to make the `exchangeData` object as generic as possible, then provide some examples of what the object may look like using bbs/zcaps/payments/etc as you listed above.
> 
> Any issues with this plan?

No, other than that those examples aren't "fully formed". We will want to improve them, which will mean we should either start with an empty object example where we put an ellipsis inside of it with a comment like "any exchange-specific data" or something -- or we just have some semi-bogus examples that we will need to iterate on. I wouldn't want to hold up the PR waiting to get good examples.

**Comment by eric-schuh** [comment]
https://github.com/w3c/vcalm/pull/713 has been raised to address this.
